clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1425,12 +1427,23 @@ public long position(TopicPartition partition) {
             if (!this.subscriptions.isAssigned(partition))
                 throw new IllegalArgumentException("You can only check the position for partitions assigned to this consumer.");
             Long offset = this.subscriptions.position(partition);
-            while (offset == null) {
+            final long startMs = time.milliseconds();
+            long finishMs = time.milliseconds();
+            while (offset == null && finishMs - startMs < requestTimeoutMs) {
                 // batch update fetch positions for any partitions without a valid position
                 updateFetchPositions();
```

Comment:
I think this method can still block regardless of requestTimoutMs at least due to the following calls:
```
updateFetchPositions() ->

coordinator.refreshCommittedOffsetsIfNeeded() ->

fetchCommittedOffsets(missingFetchPositions) ->

ensureCoordinatorReady() {
  // Using zero as current time since timeout is effectively infinite
  ensureCoordinatorReady(0, timeoutMs = Long.MAX_VALUE)
}```

Reply:
This is somewhat contradictory from what @hachikuji  said. This operation was performed asynchronously. You could take a look at conversation above for reference.

Reply:
To clarify, there are two steps in `updateFetchPositions`. The first is fetching committed offsets; the second is lookup up missing positions. The latter is asynchronous now, but the former still blocks. We'll need to fix this once the KIP is approved, but it could turn out to be a little tricky.